pmlc/conversion/tile_to_pxa/tests/compare.mlir

```diff
@@ -0,0 +1,201 @@
+  return %1, %2, %3, %4, %5, %6 : !bool, !bool, !bool, !bool, !bool, !bool
+}
+
+// -----
+
+!fp32 = type tensor<!eltwise.fp32>
+!i32  = type tensor<!eltwise.i32>
+!bool = type tensor<!eltwise.bool>
+
+// CHECK-LABEL: func @cmp_i32_fp32
+func @cmp_i32_fp32(%i32: !i32) -> (!bool, !bool, !bool, !bool, !bool, !bool) {
+  %0 = "eltwise.sconst"() {value = 0.0 : f32} : () -> !fp32
+  %1 = "eltwise.cmp_eq"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "oeq"
+  %2 = "eltwise.cmp_ne"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "one"
+  %3 = "eltwise.cmp_lt"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "olt"
+  %4 = "eltwise.cmp_le"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "ole"
+  %5 = "eltwise.cmp_gt"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "ogt"
+  %6 = "eltwise.cmp_ge"(%i32, %0) {type = !eltwise.fp32} : (!i32, !fp32) -> !bool
+// CHECK: pxa.parallel_for
+// CHECK: sitofp
+// CHECK: cmpf "oge"
+  return %1, %2, %3, %4, %5, %6 : !bool, !bool, !bool, !bool, !bool, !bool
+}
+
+
+// -----
+
+!i16 = type tensor<!eltwise.i16>
+!bool = type tensor<!eltwise.bool>
+
+// CHECK-LABEL: func @cmp_i16_i16
+func @cmp_i16_i16(%i16: !i16) -> !bool {
+  %c0 = "eltwise.sconst"() {value = 0 : i64} : () -> !i16
+  %0 = "eltwise.cmp_lt"(%i16, %c0) {type = !eltwise.fp32} : (!i16, !i16) -> !bool
+  // CHECK: pxa.parallel_for
+  // CHECK: cmpi "slt"
+  return %0 : !bool
+}
+
+// -----
+
+!i16 = type tensor<!eltwise.i16>
+!u16 = type tensor<!eltwise.u16>
+!bool = type tensor<!eltwise.bool>
+
+// CHECK-LABEL: func @cmp_i16_u16
+func @cmp_i16_u16(%i16: !i16) -> !bool {
+  %c0 = "eltwise.sconst"() {value = 0 : i64} : () -> !u16
+  %0 = "eltwise.cmp_lt"(%i16, %c0) {type = !eltwise.fp32} : (!i16, !u16) -> !bool
+  // CHECK: pxa.parallel_for
+  // CHECK: cmpi "ult"
+  return %0 : !bool
+}
+
+// -----
+
+!i16 = type tensor<!eltwise.i16>
+!i32 = type tensor<!eltwise.i32>
+!bool = type tensor<!eltwise.bool>
+
+// CHECK-LABEL: func @cmp_i16_i32
+func @cmp_i16_i32(%i16: !i16) -> !bool {
+  %c0 = "eltwise.sconst"() {value = 0 : i64} : () -> !i32
+  %0 = "eltwise.cmp_lt"(%i16, %c0) {type = !eltwise.fp32} : (!i16, !i32) -> !bool
+  // CHECK: pxa.parallel_for
+  // CHECK: sexti
+  // CHECK: cmpi "slt"
+  return %0 : !bool
+}
+
+// -----
+
+!i16 = type tensor<!eltwise.i16>
+!u32 = type tensor<!eltwise.u32>
+!bool = type tensor<!eltwise.bool>
+
+// CHECK-LABEL: func @cmp_i16_u32
+func @cmp_i16_u32(%i16: !i16) -> !bool {
+  %c0 = "eltwise.sconst"() {value = 0 : i64} : () -> !u32
+  %0 = "eltwise.cmp_lt"(%i16, %c0) {type = !eltwise.fp32} : (!i16, !u32) -> !bool
+  // CHECK: pxa.parallel_for
+  // CHECK: zexti
```

Comment:
I think this should be sext, but I'm not sure.  Might we worth checking with via compilation.  Specifically, we could check if int16_t(-1) > uint32_t(0x100000).  If it's sign extension then upcast, we should go 0xffff -> 0xffffffff which is bigger than 0x100000 when compared as unsigned.  Otherwise, if would be 0xffff -> 0x0000ffff which would be smaller than 0x100000 when compared as unsigned.

Reply:
Actually I got curious on this (and the comment later), so I just wrote some c test code:

#include <cstdint>
#include <stdio.h>

int main() {
  uint16_t a = 0xffff;
  int16_t b = -1;
  int32_t z = 0;
  uint32_t big = 0x100000;
  printf("Case 1: %d\n", a < z);
  printf("Case 2: %d\n", b > big);
}

The results are:

Case 1: 0
Case 2: 1

So definitely sext here.